### DISCUSSION TITLE
How do you open directory on android so user can select a file and upload?

### DISCUSSION THREAD

**Discussion by zburlott4** [airsdk/Adobe-Runtime-Support#3360] [General]
How do you open directory on android so user can select a file and upload? The following code works on windows no problem. I've added 

               <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
               <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />

to the android manifest. What else could possibly be missing. This was tried on SDK 50.2.4.5. and android version 14 on a samsung tablet. This should be basic functionality and is important that it works. 

```as3
_file = new File();
            
_file.addEventListener(PermissionEvent.PERMISSION_STATUS, permissionStatus);
_file.requestPermission();

function permissionStatus(event:PermissionEvent):void
{
   _file.removeEventListener(PermissionEvent.PERMISSION_STATUS, permissionStatus );
   
   if (event.status == PermissionStatus.DENIED || event.status == PermissionStatus.UNKNOWN) {
      NDAlert.show("Unable to access files");
   } else {
      _file.addEventListener(Event.SELECT, fileSelected);
      
      try
      {
         _file.browseForOpen("Open");
      }
      catch (error2:Error)
      {
      }
   }
   
   
   
}

function fileSelected(event:Event):void 
{ 
   
   _file.removeEventListener(Event.SELECT, fileSelected);
   
   
   var urlVariables :URLVariables = new URLVariables( );
   urlVariables.a= data;

   
   var urlRequest :URLRequest = new URLRequest( );
   urlRequest.method = URLRequestMethod.POST;
   urlRequest.data = urlVariables;
   

   var url:String = "<file transfer url>";
   
   //set names and URL
   urlRequest.url = url;
   
   
   //Upload
   _file.addEventListener( DataEvent.UPLOAD_COMPLETE_DATA, onUploadComplete ); 
   _file.addEventListener( IOErrorEvent.IO_ERROR, onUploadFail ); 
   _file.addEventListener( SecurityErrorEvent.SECURITY_ERROR, onUploadFail ); 
   
   _file.upload( urlRequest, "file");
   
}
```

**Comment by ajwfrost** [comment]
This is one of those side-effects of Android changing their policy on file accesses and URI schemas... if you select something, we don't now get a file path that we can pass to the system to upload it. We've just been looking at this one and can work around it using an additional method to pull out the chosen file after it's got the permission, store it temporarily locally and then delete it once the upload thread completes.

Fix will be in AIR 51.1 as it's changing the APIs between the C++ and Java side...

thanks

**Reply by zburlott4** [reply]
Is there a timeline for when AIR 51.1 may be released?

**Reply by ajwfrost** [reply]
Thursday or Friday now I'm afraid, it was meant to be the start of this week but a few last-minute challenges :-(

**Reply by zburlott4** [reply]
In 51.1.1.2 the file is being uploaded on Android, which is great progress, but the file ext and the file type is being lost which is a bit problematic. For example, an .jpg file was uploaded, but in remote uploads directory I'm getting <filename>.tmp. Changing the file ext to .jpg on the remote uploads directory then allowed in the image to be viewed.  On the android side, in the uploads complete handler, file exentions and file type are null, using the _file.type and _file.ext. Is there anyway to preserve the file type and ext during this process?
